core/src/tracker.ts

```diff
@@ -1,69 +1,91 @@
         notifier('loading', {
             content: 'Transaction is preparing to be mined...',
             key: messageId,
             duration: 0,
         });
     }
-    await new Promise((resolve, reject): void => {
-        // check if initial request is not rejected by the user
-        transactionPromise.catch((error: Error) => {
-            reject(error);
-        });
-        // track transaction progress
-        transactionManager.listen(transactionPromise, {
-            pending: () => {
-                if (notifier) {
-                    notifier('loading', {
-                        content: 'Transaction is pending to be mined...',
-                        key: messageId,
-                        duration: 0,
-                    });
-                }
-            },
-            mined: (tx: any) => {
-                if (notifier) {
-                    notifier('loading', {
-                        content: `Transaction was mined into block "${tx?.receipt?.blockNumber}", confirming...`,
-                        key: messageId,
-                        duration: 0,
-                    });
-                }
-            },
-            confirmed: (tx: any) => {
-                if (notifier) {
-                    notifier('success', {
-                        content: `Transaction was confirmed in block "${tx?.receipt?.blockNumber}"`,
-                        key: messageId,
-                        duration: 3,
-                    });
-                }
-                resolve(tx);
-            },
-            error: (tx: any) => {
-                const errorMessage = tx?.error?.message || 'unknown';
-                if (notifier) {
-                    notifier('error', {
-                        content: `Transaction was rejected with error: "${tx?.error?.message}"`,
-                        key: messageId,
-                        duration: 3,
-                    });
-                }
-                reject(new Error(errorMessage));
-            },
-        });
-        transactionManager.confirm(transactionPromise);
-    });
-    const transaction = await transactionPromise;
-    return transaction.hash;
+
+    // Wait for the user approval
+    let transaction: ethers.ContractTransaction;
+    try {
+        transaction = await transactionPromise;
+        if (notifier) {
+            notifier('loading', {
+                content: 'Transaction is pending to be mined...',
+                key: messageId,
+                duration: 0,
+            });
+        }
+    } catch (error: any) {
+        if (notifier) {
+            notifier('error', {
+                content: `Transaction error: ${truncateText(error?.message || 'unknown')}`,
+                key: messageId,
+                duration: DEFAULT_NOTIFICATION_DURATION,
+            });
+        }
+        throw new Error(error);
+    }
+
+    try {
+        // Wait for mining
+        const minedTransactionReceipt = await transaction.wait(1);
```

Comment:
Is `wait(1)` enough? I remember that sometimes a transaction will not make it into the next block (e.g., if the estimated gas price was off).

Reply:
There are two `.wait()` calls in order to better notify the user: this one is used to tell the user that transaction succeeded, while the next one below is validating 5 blocks on top 🙂